src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -82,19 +82,20 @@ public class ModelInterpreter implements Serializable {
                 // We save the caught error, if any, for throwing at the end of the build.
                 nodeOrDockerOrNone(root.agent) {
                     toolsBlock(root.agent, root.tools) {
-                            // If we have an agent and script.scm isn't null, run checkout scm
-                            if (root.agent.hasAgent() && Utils.hasScmContext(script)) {
+                        // If we have an agent and script.scm isn't null, run checkout scm
+                        if (root.agent.hasAgent() && Utils.hasScmContext(script)) {
                                 script.checkout script.scm
-                            }
+                        }
 
-                            for (int i = 0; i < root.stages.getStages().size(); i++) {
-                                Stage thisStage = root.stages.getStages().get(i)
+                        for (int i = 0; i < root.stages.getStages().size(); i++) {
+                            Stage thisStage = root.stages.getStages().get(i)
 
-                                script.stage(thisStage.name) {
-                                    if (firstError == null) {
+                            script.stage(thisStage.name) {
+                                if (firstError == null) {
+                                    nodeOrDockerOrNone(thisStage.agent) {
                                         try {
                                             catchRequiredContextForNode(root.agent) {
```

Comment:
Yes, but you should be requiring the context for the node you are working on right? If a step requests the current workspace to be injected in this context it would get the node for the stage, not the outer one. And if it is something wrong you would check the context of the wrong agent.